packages/protocol/src/interaction/SubscriptionClient.ts

```diff
@@ -0,0 +1,107 @@
+/**
+ * @license
+ * Copyright 2022-2025 Matter.js Authors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+import { MessageExchange } from "#protocol/MessageExchange.js";
+import { ProtocolHandler } from "#protocol/ProtocolHandler.js";
+import { Environment, Environmental, Logger, MaybePromise, Time, Timer } from "@matter/general";
+import { INTERACTION_PROTOCOL_ID } from "@matter/types";
+import { DataReport, IncomingInteractionClientMessenger } from "./InteractionMessenger.js";
+
+const logger = Logger.get("SubscriptionClient");
+
+export interface RegisteredSubscription {
+    id: number;
+    maximumPeerResponseTime: number;
+    maxIntervalS: number;
+    onData: (dataReport: DataReport) => MaybePromise<void>;
+    onTimeout?: () => void;
+}
+
+/**
+ * A simple protocol handler that handles exchanges starting with data reports.
+ *
+ * Incoming data reports must match to a subscription registered with {@link add} or the exchange is invalid.
+ */
+export class SubscriptionClient implements ProtocolHandler {
+    readonly #listeners = new Map<number, (dataReport: DataReport) => MaybePromise<void>>();
+    readonly #timeouts = new Map<number, Timer>();
+
+    constructor() {}
+
+    static [Environmental.create](env: Environment) {
+        const client = new SubscriptionClient();
+        env.set(SubscriptionClient, client);
+        return client;
+    }
+
+    readonly id = INTERACTION_PROTOCOL_ID;
+
+    /**
+     * Register a subscription.
+     */
+    add(subscription: RegisteredSubscription) {
+        const { id, onData, onTimeout } = subscription;
+
+        this.#listeners.set(id, onData);
+        if (onTimeout) {
+            let timer = this.#timeouts.get(id);
+            if (timer !== undefined) {
+                timer.stop();
+                this.#timeouts.delete(id);
+            }
+
+            const maxIntervalMs = subscription.maxIntervalS * 1000 + subscription.maximumPeerResponseTime;
+
+            timer = Time.getTimer("Subscription timeout", maxIntervalMs, () => {
+                logger.info(`Subscription ${id} timed out after ${maxIntervalMs}ms`);
+                this.delete(id);
+                onTimeout();
+            }).start();
+
+            this.#timeouts.set(id, timer);
+        }
+    }
+
+    /**
+     * Unregister a subscription.
+     */
+    delete(id: number) {
+        this.#listeners.delete(id);
+        const timer = this.#timeouts.get(id);
+        if (timer !== undefined) {
+            timer.stop();
+            this.#timeouts.delete(id);
+        }
+    }
+
+    async onNewExchange(exchange: MessageExchange) {
```

Comment:
do you miss the initial message here that is send in when coming from InteractionServer.call? (Or i missed something in the code)

Reply:
You mean when it delegates data report messages?  Looks OK to me but I haven't tested that wiring so could be that I got something wrong

Reply:
yes The delegationcode calls onNewExchange(exchange, message)

Reply:
WOuld be cool if you could look , else it is again par of unused yet codeflow

Reply:
Oh, I see.  I provided the message because it's part of the type signature for `onNewExchange` now.  `InteractionServer` needs it to delegate to `clientHandler` but `SubscriptionClient` doesn't use it.

Reply:
LGTM after checking deeper